File: Data Structures & Algorithms/word-break/submission-5.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        cache = {len(s):True}
        def dfs(i):
            if i in cache:
                return cache[i]

            for w in wordDict:
                l = len(w)
                if s[i:(i+l)] == w:
                    res = dfs(i+l)
                    if res:
                        cache[i] = True
                        return True
                    cache[i] = False
            return False
        
        return dfs(0)

        

File: Data Structures & Algorithms/count-connected-components/submission-5.cpp
class Solution {
public:
    unordered_map<int,vector<int>> adj;
    unordered_set<int> visited;
    int res=0;

    bool dfs(int curr, vector<vector<int>>& edges) {
        if (visited.contains(curr)) return false;
        visited.insert(curr);

        for (int neighbor : adj[curr]) {
            if (visited.contains(neighbor)) continue;
            dfs(neighbor,edges);
        }
        return true;
    }

    int countComponents(int n, vector<vector<int>>& edges) {
        for (const auto& edge : edges) {
            int a = edge[0];
            int b = edge[1];
            adj[a].push_back(b);
            adj[b].push_back(a);
        }

        for (int i=0; i<n; i++) if (dfs(i,edges)) res++;
        return res;
    }
};
